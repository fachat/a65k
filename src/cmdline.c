
/****************************************************************************

    commandline parsing
    Copyright (C) 2016-2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdlib.h>
#include <strings.h>
#include <string.h>

#include "hashmap.h"
#include "array_list.h"
#include "cmdline.h"
#include "err.h"
#include "infiles.h"

// hash from param name to cmdline_t sruct for quick find
static hash_t *params = NULL;
// list of parameters for help output
static list_t *paramlist = NULL;

static const char *prg_name = NULL;

static err_t usage(int flag, void* extra) {
	printf("Usage: %s [options] in-filename1 [in-filename2 ...]\n"
		"Cross-assembler for 65xx/R65C02/65816/65CE02/65002\n"
		"\n", prg_name); 

	list_iterator_t *iter = list_iterator(paramlist);
	cmdline_t *param = NULL;
	while ( (param = list_iterator_next(iter)) ) {
		printf("  -%s\n\t%s\n", param->name, param->description);
	}
	
	return E_ABORT;
}

static const cmdline_t help[] = {
	{ "?", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
	{ "help", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
};

static const char *key_from_param(const void *entry) {
	return ((cmdline_t*)entry)->name;
}

// TODO init the cmdline parser with the name as which the program was called
// for example "a65k", but also "xa65"
void cmdline_module_init() {

	params = hash_init_stringkey(50, 25, key_from_param);
	paramlist = array_list_init(20);

	cmdline_register_mult(help, sizeof(help)/sizeof(cmdline_t));
}

// template method where extra_param is the pointer to an int variable to set
err_t cmdline_set_flag(int flag, void *extra_param) {
	*((int*)extra_param) = flag;
	return E_OK;
}

void cmdline_register_mult(const cmdline_t *param, int num ) {
	for (int i = 0; i < num; i++) {
		cmdline_register(&param[i]);
	}
}

void cmdline_register(const cmdline_t *param) {
	
	if (hash_put(params, (void*)param) != NULL) {
		// must not happen
		exit(1);
	}
	list_add(paramlist, (void*)param);
}

err_t cmdline_parse(int argc, char *argv[]) {

	err_t rv = E_OK;

	prg_name = argv[0];

        int i = 1;
        while (i < argc && !rv) {
                if (argv[i][0] == '-') {
			char *val = NULL;
			int flag = 1;
			const char *name = argv[i]+1;
			if (name == strstr(name, "no-")) {
				name = name+3;
				flag = 0;
			}
			char *end = index(name, '=');
			if (end != NULL) {
				val = end + 1;
				end[0] = 0;
			}
			
			// lookup option
			cmdline_t *opt = hash_get(params, name);
			if (opt != NULL) {
				switch (opt->type) {
				case PARTYPE_FLAG:
					rv = opt->setflag(flag, opt->extra_param);
					break;
				case PARTYPE_PARAM:
					// TODO: error if param is missing
					// TODO: if started with "no-" return error
					rv = opt->setfunc(val, opt->extra_param);
					break;
				}
			} else {
				// TODO error cmdline parameter not found
			}
                } else {
			// register file with snapshot (clone) of current parser configuration
                        infiles_register(argv[i], parser_current_config());
                }   
		i++;
        }   
	return rv;
}




