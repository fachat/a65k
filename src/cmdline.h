
/****************************************************************************

    commandline parsing
    Copyright (C) 2016 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef CMDLINE_H
#define CMDLINE_H

#include "err.h"

// init the cmdline parser 
void cmdline_module_init();

// parse the options
err_t cmdline_parse(int argc, char *argv[]);

typedef enum {
	// no parameter, sets a bool when present with/without "no-" prefix
	PARTYPE_FLAG,
	// with required parameter
	PARTYPE_PARAM
} param_type_t;

typedef struct {
	// name of cmdline param
	const char 	*name;		

	// option has a parameter
	param_type_t	type;

	// function to call when parsed and type is PARTYPE_PARAM
	err_t 		(*setfunc)(const char *value, void *extra_param);

	// function to call when parsed and type is PARTYPE_FLAG
	void 		(*setflag)(int flag, void *extra_param);

	// extra param to be passed to set function
	void		*extra_param;

	// description
	const char	*description;
} cmdline_t;

// template method where extra_param is the pointer to an int variable to set
void cmdline_set_flag(int flag, void *extra_param);
 
void cmdline_register(const cmdline_t *param);

#endif

