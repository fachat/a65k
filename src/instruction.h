
/****************************************************************************

    CPU management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef OPERATION_H
#define OPERATION_H



typedef enum {
	SY_IMP		= 0,	// implied
	SY_IMM		= 1,	// immediate
	
	SY_ABS		= 2,	// absolute/relative address: zp; word; long; quad; rel; relwide; rellong; relquad
	SY_ABSX		= 3,	// absolute x-indexed: zp,x; word,x; long,x; quad,x
	SY_ABSY		= 4,	// absolute y-indexed: zp,y; word,y; long,y; quad,y
	SY_ABSZ		= 5,	// absolute y-indexed: zp,y; word,y; long,y; quad,y

	SY_IND		= 6,	// zp/word-indirect, word address: (zp); (word)
	SY_INDY		= 7,	// zp/word-indirect, y-indexed, word address: (zp),y; (word),y
	SY_INDZ		= 8,	// zp/word-indirect, y-indexed, word address: (zp),y; (word),y
	SY_XIND		= 9,	// x-indexed, zp/word-indirect, word address: (zp,x); (word,x)

	SY_INDL		= 10,	// zp/word-indirect, long address: [zp]; [word]
	SY_INDYL	= 11,	// zp/word-indirect, y-indexed, long address: [zp],y; [word],y
	SY_INDZL	= 12,	// zp/word-indirect, y-indexed, long address: [zp],y; [word],y
	SY_XINDL	= 13,	// x-indexed, zp/word-indirect, long address: [zp,x]; [word,x]

	SY_INDQ		= 14,	// zp/word-indirect, quad address: [[zp]]; [[word]]
	SY_INDYQ	= 15,	// zp/word-indirect, y-indexed, quad address: [[zp]],y; [[word]],y
	SY_INDZQ	= 16,	// zp/word-indirect, y-indexed, quad address: [[zp]],y; [[word]],y
	SY_XINDQ	= 17,	// x-indexed, zp/word-indirect, quad address: [[zp,x]]; [[word,x]]
	
	SY_MV		= 18,	// two byte 65816 MVN/MVP addressing
	SY_BBREL	= 19,	// zeropage address, plus relative address, R65C02 BBR/BBS
	
	SY_MAX		= 20,	// end

	// virtual addressing modes. Will be mapped to SY_IND, SY_INDL or SY_INDQ depending
	// on memory model used during assembly
	SY_INDD		= 21,	// zp/word-indirect, dynamic address: ((zp)); ((word))
	SY_INDYD	= 22,	// zp/word-zp-indirect, y-indexed, dynamic address: ((zp)),y; ((word)),y
	SY_INDZD	= 23,	// zp/word-zp-indirect, y-indexed, dynamic address: ((zp)),y; ((word)),y
	SY_XINDD	= 24,	// x-indexed, zp/word-indirect, dynamic address: ((zp,x)); ((word,x))

} syntax_type;


typedef enum {
	AM_NONE		= -1,

	AM_IMP		= 0,	// implied
	AM_IMM8		= 1,	// immediate
	AM_IMM16	= 2,	// immediate
	AM_IMM32	= 3,	// immediate
	AM_IMM64	= 4,	// immediate
	
	AM_ABS8		= 5,	// absolute/relative address: zp
	AM_ABS16	= 6,	// absolute/relative address: word
	AM_ABS24	= 7,	// absolute address: word+bank byte
	AM_ABS32	= 8,	// absolute/relative address: long
	AM_ABS64	= 9,	// absolute/relative address: quad

	AM_ABS8X	= 10,	// absolute x-indexed: zp,x
	AM_ABS16X	= 11,	// absolute x-indexed: word,x
	AM_ABS32X	= 12,	// absolute x-indexed: long,x
	AM_ABS64X	= 13,	// absolute x-indexed: quad,x

	AM_ABS8Y	= 14,	// absolute y-indexed: zp,y
	AM_ABS16Y	= 15,	// absolute y-indexed: word,y
	AM_ABS32Y	= 16,	// absolute y-indexed: long,y
	AM_ABS64Y	= 17,	// absolute y-indexed: quad,y

	AM_IND8		= 18,	// zp/word-indirect, word address: (zp)
	AM_IND16	= 19,	// zp/word-indirect, word address: (word)
	AM_IND8Y	= 20,	// zp/word-indirect, y-indexed, word address: (zp),y
	AM_IND16Y	= 21,	// zp/word-indirect, y-indexed, word address: (word),y
	AM_IND8Z	= 22,	// zp/word-indirect, z-indexed, word address: (zp),z
	AM_IND16Z	= 23,	// zp/word-indirect, z-indexed, word address: (word),z
	AM_XIND8	= 24,	// x-indexed, zp/word-indirect, word address: (zp,x)
	AM_XIND16	= 25,	// x-indexed, zp/word-indirect, word address: (word,x)

	AM_IND8L	= 26,	// zp/word-indirect, long address: [zp]
	AM_IND16L	= 27,	// zp/word-indirect, long address: [word]
	AM_IND8YL	= 28,	// zp/word-indirect, y-indexed, long address: [zp],y
	AM_IND16YL	= 29,	// zp/word-indirect, y-indexed, long address: [word],y
	AM_IND8ZL	= 30,	// zp/word-indirect, z-indexed, long address: [zp],z
	AM_IND16ZL	= 31,	// zp/word-indirect, z-indexed, long address: [word],z
	AM_XIND8L	= 32,	// x-indexed, zp/word-indirect, long address: [zp,x]
	AM_XIND16L	= 33,	// x-indexed, zp/word-indirect, long address: [word,x]

	AM_IND8Q	= 34,	// zp/word-indirect, quad address: [[zp]]
	AM_IND16Q	= 35,	// zp/word-indirect, quad address: [[word]]
	AM_IND8YQ	= 36,	// zp/word-indirect, y-indexed, quad address: [[zp]],y
	AM_IND16YQ	= 37,	// zp/word-indirect, y-indexed, quad address: [[word]],y
	AM_IND8ZQ	= 38,	// zp/word-indirect, z-indexed, quad address: [[zp]],z
	AM_IND16ZQ	= 39,	// zp/word-indirect, z-indexed, quad address: [[word]],z
	AM_XIND8Q	= 40,	// x-indexed, zp/word-indirect, quad address: [[zp,x]]
	AM_XIND16Q	= 41,	// x-indexed, zp/word-indirect, quad address: [[word,x]]
	
	AM_MV		= 42,	// two byte 65816 MVN/MVP addressing
	AM_BBREL	= 43,	// zeropage address, plus relative address, R65C02 BBR/BBS
	
	AM_MAX		= 44,	// end
} amode_type;
	
typedef enum {
	PG_BASE		= 0,	// base page, no page prefix needed
	PG_EXT		= 1,	// extended page
	PG_QUICK	= 2,	// quick page
	PG_SYS		= 3,	// system page
} pg_type;

typedef enum {
	RS_NOT_SET	= 0,	// not set
        RS_BYTE         = 1,    // byte width 
        RS_WORD         = 2,    // word (2 byte) width
        RS_LONG         = 3,    // long (4 byte) width
        RS_QUAD         = 4     // quad (8 byte) width
} rs_type;

typedef enum {
	LE_NOT_SET	= 0,	// not set
        LE_E            = 1,    // no extension
        LE_S            = 2,    // sign extensions
        LE_0            = 3,    // extend with 0
        LE_1            = 4     // extend with 1
} le_type;


// ordering by SY_*, then AM_*
#define AMODES_MAX      5

typedef struct {
	const char	*pre;
	const char	*post;
	amode_type	amode[AMODES_MAX];
} syntax_t;

typedef struct {
	//bool_t		is_valid;
	unsigned char	code;
	pg_type 	page_prefix;
	const char 	*amodename;
	isa_map		isa;
} opcode_t;

// operation - equivalent to the mnemonic, like "lda", "adc", "inx", ...
typedef struct instruction_s instruction_t;

struct instruction_s {
	const char	*name;
	isa_map		isa;		// what ISA is it from? Also selects BCD and Illegal opcodes
	bool_t		abs_is_rel;	// is branch with relative addressing?
	bool_t		check_ac_w;	// check AC width mode for 65816, for adr modes with check_width
	bool_t		check_idx_w;	// check Index register width mode for 65816 (with check_width)
	instruction_t	*next;
	opcode_t	opcodes[AM_MAX];
};

typedef struct {
	bool_t		set_am_prefix;
	bool_t		rs_is_width;		// use RS-prefix for width
	pg_type		page_prefix;
	unsigned char	code;
} codepoint_t;

void instruction_module_init();

const instruction_t *instruction_find(const char *name);

bool_t opcode_find(const position_t *loc, const context_t *ctx, const instruction_t *op, syntax_type syntax, int opsize_in_bytes,
	codepoint_t *returned_code);

const syntax_t *op_syn_details(syntax_type syn);

inline static char prefix_um_char(bool_t um) {
	return um ? 'U' : '-';
}

inline static char prefix_nf_char(bool_t nf) {
	return nf ? 'N' : '-';
}

inline static char prefix_rs_char(rs_type p) {
	switch (p) {
	case RS_NOT_SET:
		return '-';
	case RS_BYTE:
		return 'B';
	case RS_WORD:
		return 'W';
	case RS_LONG:
		return 'L';
	case RS_QUAD:
		return 'Q';
	}
	return 'x';
}

inline static char prefix_le_char(rs_type p) {
	switch (p) {
	case LE_NOT_SET:
		return '-';
	case LE_0:
		return '0';
	case LE_1:
		return '1';
	case LE_S:
		return 'S';
	case LE_E:
		return 'E';
	}
	return 'x';
}

#endif

