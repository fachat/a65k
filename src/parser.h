/****************************************************************************

    parser
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PARSER_H
#define PARSER_H

#include "context.h"
#include "label.h"
#include "block.h"
#include "operation.h"
#include "tokenizer.h"
#include "arith.h"
#include "list.h"

typedef enum {
        S_LABEQPC,              // set label to PC
        S_LABDEF,               // set label from parameter
	S_OPCODE,		// opcode
	S_PSEUDO,		// pseudo opcode
} stype_t;

/**
 * The statement_t struct is the main structure of the internal workings.
 * The parser creates a statement_t with all information to be gathered from the source.
 * The printer can "lint" this statement and pretty-format it
 * The assembler itself resolves labels and computes references and extends it with binary values
 * The printer can then also print the extended statement with binary values
 * The renderer can then put (a list of) statements into a binary format, depending on the output options
 */
typedef struct {
        const block_t           *blk;		// pointer to current block structure for label dereferencing
        const context_t         *ctx;		// current context, like CPU type, width, other options
        stype_t                 type;		// type of statement (see above)
        // optional
        const label_t           *label;		// TODO allow for multiple label definitions per line/statement
	// S_LABDEF
	const ilist_t		*setlabel;	// anode_t list for setting label to a value 
	// S_OPCODE
       	const operation_t       *op;		// pointer to info about operation, like "ADC", "TAX", ...
	op_t			base;		// base extension of the parameter, like ",B", ",s" if any, zero else
	
	syntax_type		syn;		// addressing mode syntax, see SY_* constants
       	ilist_t  	        *param;		// list of anode_t structs that describe the syntax tree for the parameter value
} statement_t;


void parser_module_init(void);

err_t parser_push(const context_t *context, const line_t *line);

list_iterator_t *parser_get_statements(void);

// for test cases only
void parser_reset();

#endif

