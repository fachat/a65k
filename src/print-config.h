/****************************************************************************

    printer config
    Copyright (C) 2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PRINT_CONFIG_H
#define PRINT_CONFIG_H


typedef struct {
	// output filename
	const char 	*filename;
	// print the line number, or reserve space for it if not parsed
	int 		lineno;
	// if true (default) collapse multiple empty lines into a single empty line
	int 		collapse_empty;
} print_config_t;

void print_config_init(); 

// return a clone of the current configuration
print_config_t *print_current_config();

#endif

