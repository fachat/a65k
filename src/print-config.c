/****************************************************************************

    printer config
    Copyright (C) 2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <string.h>

#include "mem.h"
#include "err.h"
#include "print-config.h"
#include "cmdline.h"


static type_t print_config_memtype = {
        "print_config_t",
        sizeof(print_config_t)
};

static print_config_t pconfig;


static cmdline_t print_params[] = {
	{ "print-lineno", NULL, PARTYPE_FLAG, NULL, cmdline_set_flag, &pconfig.lineno, 
		"print (or leave space for) line numbers", NULL },
};

void print_config_init() {

	pconfig.lineno = 0;
	
	int n = sizeof(print_params) / sizeof(cmdline_t);
	cmdline_register_mult(print_params, n);
}

print_config_t* print_current_config() {

	print_config_t *clonedconf = mem_alloc(&print_config_memtype);

	memcpy(clonedconf, &pconfig, sizeof(pconfig));

	return clonedconf;
}

