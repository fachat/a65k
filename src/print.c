/****************************************************************************

    print output
    Copyright (C) 2012,2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdarg.h>
#include <ctype.h>
#include <string.h>
#include <errno.h>

#include "print-config.h"
#include "print.h"
#include "arith.h"

// filler spaces
static const char spaces[] = "                                        ";

// column widths for fixed width output, index corresponds to printer_col_t
static const int widths[] = { 
	6, 	// PRT_LINENO
	5,	// PRT_ADDR
	9,	// PRT_DATA
	11,	// PRT_LABEL
	4,	// PRT_OPERATION
	20,	// PRT_PARAM
	30	// PRT_COMMENT
};

void print_module_init() {

	print_config_init();
}

static type_t printer_memtype = {
        "printer_t",
        sizeof(printer_t)
};

// initialize printer output channel
printer_t *print_init() {

	print_config_t *cfg = print_current_config();

	const char *filename = cfg->filename;
	FILE *f = NULL;
	if (filename != NULL && strlen(filename) > 0) {
		f = fopen(filename, "w");
		if (f == NULL) {
			log_error("Could not open output file '%s' -> %s", filename, strerror(errno));
			return NULL;
		}
	}

	printer_t *prt = mem_alloc(&printer_memtype);

	prt->outfile = f;
	prt->cfg = cfg;
	prt->current_col = PRT_LINENO;

	int p = 0;
	for (int i = PRT_LINENO; i < PRT_END; i++) {
		prt->col_width[i] = p;
		p += widths[i];
	}

	print_clr(prt);

	prt->was_empty = 0;

	return prt;
}

static int print_getlen(printer_t *prt) {
	return prt->buflen;
}

static void print_setcol(printer_t *prt, int col) {
	
	int buflen = print_getlen(prt);

	int todo = col - buflen;
	int slen = strlen(spaces);

	if (todo <= 0) {
		todo = 1;
	}

	while (todo > 0) {
		int step = slen - todo;
		if (step < 0) {
			step = 0;
		}
		print(prt, prt->current_col, "%s", spaces + step);
		todo -= slen + step;
	}
}


void print(printer_t *prt, print_col_t col_no, const char *pattern, ...) {

	if (col_no != prt->current_col) {
		// target fixed width column
		if (col_no == PRT_COMMENT && prt->current_col == PRT_LINENO) {
			// shift comment to left if only comment on line
			col_no = PRT_LABEL;
		}
		int trg = prt->col_width[col_no];
		print_setcol(prt, trg);
		prt->current_col = col_no;
	}

        va_list ap;
        va_start(ap, pattern);

	int buflen = print_getlen(prt);

        int r = vsnprintf(prt->buf + buflen, BUF_LEN - buflen, pattern, ap);
        if (r < 0 || r > BUF_LEN) {
                // error
                log_error("Error printing %d\n", r);
                return;
        }
	prt->buflen += r;
}

void print_out(printer_t *prt) {

	if (prt->buflen == 0) {
		if (prt->cfg->collapse_empty && prt->was_empty) {
			return;
		}
		prt->was_empty = 1;
	} else {
		prt->was_empty = 0;
	}

	if (prt->outfile == NULL) {
        	printf("%s\n", prt->buf);
	} else {
        	fprintf(prt->outfile, "%s\n", prt->buf);
	}
	print_clr(prt);
}

void print_clr(printer_t *prt) {
	prt->buf[0] = 0;
	prt->buflen = 0;
	prt->current_col = PRT_LINENO;
}

void print_close(printer_t *prt) {

	if (prt->outfile) {
		fclose(prt->outfile);
	}
}


