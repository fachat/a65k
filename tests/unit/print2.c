

#include <stdio.h>

#include "types.h"
#include "position.h"
#include "cpu.h"
#include "segment.h"
#include "context.h"
#include "position.h"
#include "instruction.h"
#include "cmdline.h"
#include "parser.h"
#include "print-config.h"
#include "dbgprint.h"
#include "print.h"
#include "log.h"


static void test(const context_t *ctx, const print_config_t *cfg, position_t *pos, const char *txt) {

	parser_reset();

	line_t line = { txt, pos, parser_current_config() };

	do_print("");
	do_print(">>> %s", line.line);
		
	err_t err = parser_push(ctx, &line);

	if (err) {
		do_print("err=%d", err);
	}
	//log_set_level(LEV_DEBUG);

	printer_t *prt = print_init();

	list_iterator_t *iter = parser_get_statements();
	while (list_iterator_has_next(iter)) {
		statement_t *stmt = list_iterator_next(iter);
		print_formatted_stmt(prt, stmt);
	}
}


int main(int argc, char *argv[]) {

        (void)argc;
        (void)argv;

	cmdline_module_init();
	cpu_module_init();
	segment_module_init();
	operation_module_init();
	parser_module_init();
	tokenizer_module_init();

	const print_config_t *cfg = print_current_config();

	const cpu_t *cpu = cpu_by_name("nmos");

	segment_t *seg = segment_new(NULL, "test", SEG_ANY, cpu->type, true);

	const context_t *ctx = context_init(seg, cpu);

	position_t pos = { "bogusfile", 1 };

	test(ctx, cfg, &pos, "label adc.w #1+(2*123)");	
	test(ctx, cfg, &pos, "lda 0+(2*123)");	
	test(ctx, cfg, &pos, "lda 0+(2*123),s");	

	test(ctx, cfg, &pos, "lda 123,p");	
	test(ctx, cfg, &pos, "lda <123,p");	
	test(ctx, cfg, &pos, "lda >123,p");	
	test(ctx, cfg, &pos, "lda #>123,p");	

	test(ctx, cfg, &pos, "lda -123,p");	
	test(ctx, cfg, &pos, "lda <-123,p");	
	test(ctx, cfg, &pos, "lda >-~123,p");	
	test(ctx, cfg, &pos, "lda #>-123,p");	
/*

	test(ctx, cfg, &pos, "lda 123,b");	
	test(ctx, cfg, &pos, "ldx (12),y");	
	test(ctx, cfg, &pos, "lda ($12,s,x)");	
	test(ctx, cfg, &pos, "lda ($ff,s),z");	

	test(ctx, cfg, &pos, "ldx [12],y");	
	test(ctx, cfg, &pos, "lda [$12,s,x]");	
	test(ctx, cfg, &pos, "lda [12,s],y");	
	test(ctx, cfg, &pos, "lda [$ff,s],z");	

	test(ctx, cfg, &pos, "ldx [[12]],y");	
	test(ctx, cfg, &pos, "lda [[$12,s,x]]");	
	test(ctx, cfg, &pos, "lda [[12,s]],y");	
	test(ctx, cfg, &pos, "lda [[$ff,s]],z");	

	test(ctx, cfg, &pos, "ldx ((12)),y");	
	test(ctx, cfg, &pos, "lda (($12,s,x))");	
	test(ctx, cfg, &pos, "lda ((12,s)),y");	
	test(ctx, cfg, &pos, "lda (($ff,s)),z");	

	test(ctx, cfg, &pos, "ldx.0 123");	
	test(ctx, cfg, &pos, "lda.1u 123");	
	test(ctx, cfg, &pos, "lda.ne 123");	
	test(ctx, cfg, &pos, "lda.sw 123");	
	test(ctx, cfg, &pos, ";comment");	
	test(ctx, cfg, &pos, "lda 123; comment");	
	test(ctx, cfg, &pos, "lda.sw 123; comment");	
	test(ctx, cfg, &pos, "lda.un1w (($ff,s)),z;comment test");	
*/
}



