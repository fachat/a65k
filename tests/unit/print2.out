WRN:
WRN:>>> label adc.w #1+(2*123)
                    label      ADC.W #1+(2*123)
WRN:
WRN:>>> lda 0+(2*123)
                               LDA 0+(2*123)
WRN:
WRN:>>> lda 0+(2*123),s
                               LDA 0+(2*123),S
WRN:
WRN:>>> lda 123,p
                               LDA 123,P
WRN:
WRN:>>> lda <123,p
                               LDA <123,P
WRN:
WRN:>>> lda >123,p
                               LDA >123,P
WRN:
WRN:>>> lda #>123,p
ERR:bogusfile:1 Syntax error: Expecting comment or colon
ERR:lda #>123,p
ERR:         ^
WRN:err=1
WRN:
WRN:>>> lda -123,p
                               LDA -123,P
WRN:
WRN:>>> lda <-123,p
                               LDA <-123,P
WRN:
WRN:>>> lda >-~123,p
                               LDA >-~123,P
WRN:
WRN:>>> lda #>-123,p
ERR:bogusfile:1 Syntax error: Expecting comment or colon
ERR:lda #>-123,p
ERR:          ^
WRN:err=1
